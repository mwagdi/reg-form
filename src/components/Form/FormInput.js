import React, { Component } from 'react';

class FormInput extends Component{
    state = {
        input: "",
        conditions: this.props.field.conditions.map(condition => {
            return {
                title: condition.title,
                done: false
            }
        })
    }
    handleChange = e => {
        const { conditions, label } = this.props.field;
        this.setState({ input: e.target.value });
        let cArray = JSON.parse(JSON.stringify(this.state.conditions));
        cArray.forEach((c, i) => {
            c.done = conditions[i].test(e.target.value);
        });
        this.setState({ conditions: [...cArray] }, () => {
            const done = this.state.conditions.reduce((acc, c) => {
                return acc && c.done;
            }, true);
            this.props.onChange(this.state.input, done, label);
        });
    }
    render(){
        const { label, type, name } = this.props.field;
        const id = Math.random().toString(36).substr(2, 16);
        return (
            <div className="regForm__field">
                <label htmlFor={name}>{label}</label>
                <input value={this.state.input} onChange={this.handleChange} name={name} type={type} />
                {this.state.conditions.map((c, i) => <span key={i}>{c.title} - {c.done.toString()}</span>)}
            </div>
        )
    }
}

export default FormInput;