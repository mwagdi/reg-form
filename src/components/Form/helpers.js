export const convertToKey = str => str.toLowerCase().replace(" ","_");