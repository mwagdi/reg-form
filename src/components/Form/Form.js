import React, { Component } from 'react';
import FormInput from './FormInput';
import { convertToKey } from './helpers';

class Form extends Component{
    state = this.props.fields.reduce((acc, field) => {
        return {
            ...acc,
            [convertToKey(field.label)]: {
                value: "",
                filled: false
            }
        }
    },{ valid: false })
    
    handleSubmit = e => {
        e.preventDefault();
        console.log(e);
    }
    handleChange(value, filled, field){
        this.setState({
            ...this.state,
            [convertToKey(field)]: {
                value,
                filled
            }
        }, () => {
            this.validate(); 
        })
    }
    validate(){
        this.setState({
            valid: this.props.fields.reduce((acc, f) => {
                return acc && this.state[convertToKey(f.label)].filled
            }, true)
        })
    }
    render(){
        const { title, fields } = this.props;
        return (
            <form onSubmit={this.handleSubmit} className="regForm">
                <h2 className="regForm__title">{title}</h2>
                {fields.map((field, i) => <FormInput onChange={(value, filled) => this.handleChange(value, filled, field.label)} field={field} key={i} />)}
                <input className="regForm__submit" type="submit" value="Submit"/>
            </form>
        )
    }
}

export default Form;