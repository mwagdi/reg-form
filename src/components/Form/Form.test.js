import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Form from './Form';

describe('Form', () => {
    const fields = [
        {
            label: "Email",
            type: "email"
        },
        {
            label: "Password",
            type: "password"
        }
    ]
    
    test('renders without crashing', () => {
        shallow(<Form fields={fields} />);
    });
    
    test('renders correctly', () => {
        const wrapper = shallow(<Form fields={fields} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    })
    
    test('renders form with class regForm', () => {
        const wrapper = shallow(<Form fields={fields} />);
        expect(wrapper.exists('.regForm')).toEqual(true);
    })

    test('renders FormInputs', () => {
        const wrapper = shallow(<Form fields={fields} />)
        expect(wrapper.find('FormInput').length).toEqual(fields.length);
    })

});
