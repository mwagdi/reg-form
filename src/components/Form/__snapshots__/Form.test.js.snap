// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<form
  className="regForm"
  onSubmit={[Function]}
>
  <h2
    className="regForm__title"
  />
  <FormInput
    field={
      Object {
        "label": "Email",
        "type": "email",
      }
    }
    key="0"
    onChange={[Function]}
  />
  <FormInput
    field={
      Object {
        "label": "Password",
        "type": "password",
      }
    }
    key="1"
    onChange={[Function]}
  />
  <input
    className="regForm__submit"
    type="submit"
    value="Submit"
  />
</form>
`;
